File: 2025/09 September/24 Design MinMax Queue/main.cpp
// Link: https://www.geeksforgeeks.org/problems/design-minmax-queue/1

#include <iostream>
#include <vector>
#include <queue>
#include <deque>
using namespace std;

class SpecialQueue
{
    queue<int> q1;
    deque<int> q2;

public:
    void enqueue(int x)
    {
        q1.push(x);
        while (!q2.empty() && q2.back() > x)
        {
            q2.pop_back();
        }
        q2.push_back(x);
    }

    void dequeue()
    {
        if (q1.empty())
        {
            cout << "Queue is empty\n";
            return;
        }
        int frontVal = q1.front();
        if (frontVal == q2.front())
        {
            q2.pop_front();
        }
        q1.pop();
    }

    int getMin()
    {
        if (q1.empty())
        {
            cout << "Queue is empty\n";
            return -1;
        }
        return q2.front();
    }

    int getFront()
    {
        if (q1.empty())
        {
            cout << "Queue is empty\n";
            return -1;
        }
        return q1.front();
    }

    int getMax()
    {
        if (q1.empty())
        {
            cout << "Queue is empty\n";
            return -1;
        }
        return q2.back();
    }
};

int main()
{
    SpecialQueue sq;
    sq.enqueue(4);
    sq.enqueue(2);
    cout << "Front: " << sq.getFront() << endl;
    cout << "Min: " << sq.getMin() << endl;
    sq.dequeue();
    cout << "Max: " << sq.getMax() << endl;

    SpecialQueue sq2;
    sq2.enqueue(3);
    cout << "Min: " << sq2.getMin() << endl;
    sq2.enqueue(5);
    cout << "Max: " << sq2.getMax() << endl;
    return 0;
}

File: 2025/09 September/24 Design MinMax Queue/main.py
# Link: https://www.geeksforgeeks.org/problems/design-minmax-queue/1

from collections import deque


class SpecialQueue:
    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()

    def enqueue(self, x):
        self.q1.append(x)
        while self.q2 and self.q2[-1] > x:
            self.q2.pop()
        self.q2.append(x)

    def dequeue(self):
        if not self.q1:
            print("Queue is empty")
            return
        frontVal = self.q1.popleft()
        if frontVal == self.q2[0]:
            self.q2.popleft()

    def getMin(self):
        if not self.q1:
            print("Queue is empty")
            return -1
        return self.q2[0]

    def getFront(self):
        if not self.q1:
            print("Queue is empty")
            return -1
        return self.q1[0]

    def getMax(self):
        if not self.q1:
            print("Queue is empty")
            return -1
        return max(self.q1)


if __name__ == "__main__":
    sq = SpecialQueue()
    sq.enqueue(4)
    sq.enqueue(2)
    print(f"Front element: {sq.getFront()}")
    print(f"Minimum element: {sq.getMin()}")
    sq.dequeue()
    print(f"Maximum element: {sq.getMax()}")

    sq = SpecialQueue()
    sq.enqueue(3)
    print(f"Minimum element: {sq.getMin()}")
    sq.enqueue(5)
    print(f"Maximum element: {sq.getMax()}")
